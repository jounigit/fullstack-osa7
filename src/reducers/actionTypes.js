export const INIT_BLOGS = 'INIT_BLOGS'
export const LIKE_BLOG = 'LIKE_BLOG'
export const CREATE_BLOG = 'CREATE_BLOG'
export const DELETE_BLOG = 'DELETE_BLOG'
export const ADD_COMMENT = 'ADD_COMMENT'

export const INIT_USERS = 'INIT_USERS'

export const NEW_MSG = 'NEW_MSG'
export const HIDE_MSG = 'HIDE_MSG'

export const ADD_ERROR = 'ADD_ERROR'
export const CLEAR_ERROR = 'CLEAR_ERROR'
