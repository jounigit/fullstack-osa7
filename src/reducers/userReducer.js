import usersService from '../services/users'

const userReducer = (store = [], action) => {

  if (action.type === 'INIT') {
    return action.data
  }

  return store
}

export const userInit = () => {
  return async (dispatch) => {
    const users = await usersService.getAll()
    dispatch({
      type: 'INIT',
      data: users
    })
  }
}

export default userReducer
