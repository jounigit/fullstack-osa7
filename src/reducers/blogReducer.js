import blogService from '../services/blogs'
import { INIT_BLOGS, LIKE_BLOG, CREATE_BLOG, DELETE_BLOG, NEW_MSG, ADD_COMMENT } from './actionTypes'


const blogReducer = (state = [], action) => {
  switch (action.type) {
  case INIT_BLOGS:
    return action.data
  case CREATE_BLOG:
    return [...state, action.data]
  case LIKE_BLOG:
    return state.map(b => b.id !== action.data.id ? b : action.data)
  case DELETE_BLOG:
    return state.filter(b => b.id !==action.id)
  case ADD_COMMENT:
    return state.map(b => b.id !== action.data.id ? b : action.data)
  default:
    return state
  }

}

export const initBlogs = () => async (dispatch) => {
  const blogs = await blogService.getAll()
  dispatch({
    type: INIT_BLOGS,
    data: blogs
  })
}

export const createBlog = (content) => async (dispatch) => {
  try {
    const data = await blogService.create(content)
    dispatch({
      type: CREATE_BLOG,
      data
    })
  } catch(error) {
    console.log('ERROR:: ', error.message)
    dispatch({
      type: NEW_MSG,
      content: error.message,
      style: 'error'
    })
  }
}

export const addComment = (id, commentObj) => async (dispatch) => {
  try {
    const data = await blogService.addComment(id, commentObj)
    console.log('Add comment:: ', data)
    dispatch({
      type: ADD_COMMENT,
      data
    })
  } catch(error) {
    console.log('ERROR:: ', error.message)
    dispatch({
      type: NEW_MSG,
      content: error.message,
      style: 'error'
    })
  }
}


export const deleteBlog = (id) => async (dispatch) => {
  try {
    await blogService.remove(id)
    dispatch({
      type: DELETE_BLOG,
      id
    })
  } catch(error) {
    dispatch({
      type: NEW_MSG,
      content: error.message,
      style: 'error'
    })
  }
}

export const like = (id) => async (dispatch) => {
  try {
    const blogs = await blogService.getAll()
    const blog = blogs.find(b => b.id === id)
    const toUpdate = { ...blog, likes: blog.likes+1 }
    await blogService.update(id, toUpdate)
    dispatch({
      type: LIKE_BLOG,
      data: toUpdate
    })
  } catch(error) {
    dispatch({
      type: NEW_MSG,
      content: error.message,
      style: 'error'
    })
  }
}

export default blogReducer

/*
export const createBlog = (content) => {
  return async (dispatch) => {
    try {
      const data = await blogService.create(content)
      dispatch(createBlogSuccess(data))
    } catch(e) {
      dispatch(createBlogError(e))
    }
  }
}

const createBlogSuccess = (data) => {
  return {
    type: 'CREATE_BLOG',
    data
  }
}

const createBlogError = (e) => {
  return {
    type: 'ERROR',
    e
  }
}
*/
