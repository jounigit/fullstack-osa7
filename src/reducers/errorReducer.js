import { ADD_ERROR, CLEAR_ERROR } from './actionTypes'

const initState = {
  message: null
}

const errorReducer = (state = initState, action) => {
  switch (action.type) {
  case ADD_ERROR:
    return state = { message: action.message }
  case CLEAR_ERROR:
    return state = initState
  default:
    return state
  }
}

export const errorMessage = (message, time) => (dispatch) => {
  dispatch({
    type: ADD_ERROR,
    message
  })
  setTimeout(() => {
    dispatch({
      type: CLEAR_ERROR
    })
  }, (time * 1000))
}

export default errorReducer
