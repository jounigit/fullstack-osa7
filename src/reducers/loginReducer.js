import loginService from '../services/login'
import blogService from '../services/blogs'
import { LOGIN, LOGOUT, LOGGEDIN, NEW_MSG } from './actionTypes'

export const login = (username, password) => async (dispatch) => {
  try {
    const user = await loginService.login({
      username,
      password
    })

    window.localStorage.setItem('loggedBlogAppUser', JSON.stringify(user))
    blogService.setToken(user.token)

    dispatch({
      type: LOGIN,
      user
    })
  } catch(error) {
    console.log('ERROR:: ', error.message)
    dispatch({
      type: NEW_MSG,
      content: 'wrong username or password',
      style: 'error'
    })
  }
}

export const logout = () => (dispatch) => {
  window.localStorage.removeItem('loggedBlogAppUser')
  dispatch({
    type: LOGOUT
  })
}

export const loggedUser = (user) => async (dispatch) => {
  //console.log('LOGGEDUSER:: ', user)
  dispatch({
    type: LOGGEDIN,
    user
  })
}

const loginReducer = (state = null, action) => {
  switch (action.type) {
  case LOGIN:
    return action.user
  case LOGGEDIN:
    return action.user
  case LOGOUT:
    return null
  default:
    return state
  }
}

export default loginReducer
