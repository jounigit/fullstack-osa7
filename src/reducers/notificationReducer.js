import { NEW_MSG, HIDE_MSG } from './actionTypes'
const msgAtStart = 'Message form chef'

const initialState = {
  content: msgAtStart,
  style: '',
  visibility: false
}

const notificationReducer = (state = initialState, action) => {
  switch (action.type) {
  case NEW_MSG:
    return state = { content: action.content, style: action.style, visibility: true }
  case HIDE_MSG:
    return state = initialState
  default:
    return state
  }
}

export const notify = (content, time, style) => (dispatch) => {
  dispatch({
    type: NEW_MSG,
    content,
    style
  })
  setTimeout(() => {
    dispatch({
      type: HIDE_MSG
    })
  }, (time * 1000))
}

export default notificationReducer
