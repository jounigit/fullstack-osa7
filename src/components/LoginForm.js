import React from 'react'
import { connect } from 'react-redux'
import { login } from '../reducers/loginReducer'
import { Form, Button, Header, Card } from 'semantic-ui-react'
//import PropTypes from 'prop-types'

export class LoginForm extends React.Component {
  handleSubmit = async (e) => {
    e.preventDefault()

    this.props.login(e.target.username.value, e.target.password.value)
    e.target.username.value = ''
    e.target.password.value = ''
  }
  render () {
    return (
      <Card centered style={{ marginTop: 20 }}>
        <Card.Content>
          <Header as='h2' color='green'>Kirjaudu sovellukseen</Header>
        </Card.Content>
        <Card.Content>
          <Form onSubmit={this.handleSubmit}>
            <Form.Field>
              <label>käyttäjätunnus</label>
              <input type="text" name="username" />
            </Form.Field>
            <Form.Field>
              <label>salasana</label>
              <input type="password" name="password" />
            </Form.Field>
            <Button basic color='green'>kirjaudu</Button>
          </Form>
        </Card.Content>
      </Card>
    )
  }
}
/*
LoginForm.propTypes = {
  username: PropTypes.string,
  password: PropTypes.string
} */

export default connect(
  null,
  { login }
)(LoginForm)
