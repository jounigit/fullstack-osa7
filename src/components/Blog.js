import React from 'react'
import { connect } from 'react-redux'
import { notify } from '../reducers/notificationReducer'
import Comments from './Comments'
import { Redirect } from 'react-router'
import { like, deleteBlog } from '../reducers/blogReducer'
import { Button, Header, Card } from 'semantic-ui-react'

class Blog extends React.Component {
  toggleLike = (id, content) => () => {
    this.props.like(id)
    this.props.notify(`you liked '${content}'`, 10, 'success')
  }

  remove = (id, title, author) => () => {
    console.log('Remove ', id, title)
    const ok = window.confirm(`remove blog '${title}' by ${author}?`)
    if ( ok===false) {
      return
    }
    this.props.deleteBlog(id)
    this.props.notify(`blog '${title}' by ${author} removed`, 5, 'success')
    return <Redirect to='/' />
  }

  render() {
    const { username, blog } = this.props

    if ( blog === undefined ) {
      return <Redirect to='/' />
    }
    const adder = blog.user ? blog.user.name : 'anonymous'
    const deletable = blog.user === undefined || blog.user.username === username

    return (
      <Card centered>

        <Card.Content>
          <Header as='h1'>
            {blog.title}
          </Header>
          <Header as='h3'>
            {blog.author}
          </Header>
          <a href={blog.url}>{blog.url}</a>
        </Card.Content>
        <Card.Content extra>
          <Header as='h5'>
            added by {adder}

            {deletable &&
              <Button negative  size='mini' floated='right'
                onClick={this.remove(blog.id, blog.title, blog.author)}>
                delete
              </Button>
            }
          </Header>
          <Button
            size='mini' floated='right'
            color='green'
            content='Likes'
            icon='thumbs up'
            label={{ as: 'a', basic: true, pointing: 'right', content: `${blog.likes}` }}
            labelPosition='left'
            onClick={this.toggleLike(blog.id, blog.title)}
          />
        </Card.Content>
        <Card.Content>
          <Comments blog={blog} />
        </Card.Content>
      </Card>
    )
  }
}

export default connect(
  null,
  { notify, like, deleteBlog }
)(Blog)
