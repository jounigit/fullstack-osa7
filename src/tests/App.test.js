import React from 'react'
import { shallow, mount } from 'enzyme'
import ConnectedApp, { App } from '../App'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
const mockStore = configureStore([thunk])
import { Provider } from 'react-redux'
import { login } from '../reducers/loginReducer'
import Blog from '../components/Blog'
import LoginForm from '../components/LoginForm'
import toJson from 'enzyme-to-json'

let wrapper

describe.skip('+++ Shallow Render REACT COMPONENTS', () => {
  beforeEach(() => {
    wrapper = shallow(<App />)
  })

  it('no blogs is rendered', () => {
    //app.update()
    const blogComponents = wrapper.find(Blog)
    expect(blogComponents.length).toBe(0)
  })

})

//*************************************************************
describe.skip('>>>APP --- (Shallow + passing the {store} directly)', () => {
  const initialState = { user: null }
  const mockStore = configureStore()
  let store, wrapper

  beforeEach(() => {
    store = mockStore(initialState)
    wrapper = shallow(<ConnectedApp store={store} /> )
  })

  it('+++ check Prop matches with initialState', () => {
    expect(wrapper.prop('user')).toEqual(initialState.user)
  })

  it('+++ render the connected component', () => {
    expect(wrapper.length).toEqual(1)
  })

  it('+++ renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })

})

//***************************************************************************
describe.skip('>>>APP --- (Mount + wrapping in <Provider>)', () => {
  const initialState = { user: null }
  const mockStore = configureStore()
  let store, wrapper

  beforeEach(() => {
    store = mockStore(initialState)
    wrapper = mount( <Provider store={store}><ConnectedApp /></Provider> )
  })

  it('+++ render the connected component', () => {
    expect(wrapper.find(ConnectedApp).length).toEqual(1)
  })

  it('+++ check Prop matches with initialState', () => {
    expect(wrapper.find(App).prop('user')).toEqual(initialState.user)
  })

  it('+++ check LoginForm is rendered', () => {
    const loginFormComponent = wrapper.find(App).find(LoginForm)
    console.log('LOGINFORM ---- ', loginFormComponent)
    expect(loginFormComponent.length).toBe(1)
  })

  it('+++ renders LoginForm correctly', () => {
    expect(toJson(wrapper, {
      noKey: false,
      mode: 'deep'
    })).toMatchSnapshot()
  })

})

//**************************************************************************
describe.skip('>>> APP ---- login action with wrong credentials', () => {
  it('handles login', async () => {
    const store = mockStore()
    //console.log('MOCKSTORE 2 ---- ', store)
    await store.dispatch(login('tollo', 'password'))
    const actions = store.getActions()
    expect(actions[0]).toEqual({
      type: 'NEW_MSG',
      content: 'wrong username or password',
      style: 'error'
    })
  })
})

/*
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";

export const mockStore = configureMockStore([thunk])
*/
