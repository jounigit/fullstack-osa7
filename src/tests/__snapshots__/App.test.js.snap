// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>APP --- (Mount + wrapping in <Provider>) +++ renders LoginForm correctly 1`] = `
<div
  className="ui centered card"
  onClick={[Function]}
  style={
    Object {
      "marginTop": 20,
    }
  }
>
  <div
    className="content"
  >
    <h2
      className="ui green header"
    >
      Kirjaudu sovellukseen
    </h2>
  </div>
  <div
    className="content"
  >
    <form
      className="ui form"
      onSubmit={[Function]}
    >
      <div
        className="field"
      >
        <label>
          käyttäjätunnus
        </label>
        <input
          name="username"
          type="text"
        />
      </div>
      <div
        className="field"
      >
        <label>
          salasana
        </label>
        <input
          name="password"
          type="password"
        />
      </div>
      <button
        className="ui green basic button"
        onClick={[Function]}
        role="button"
      >
        kirjaudu
      </button>
    </form>
  </div>
</div>
`;

exports[`>>>APP --- (Shallow + passing the {store} directly) +++ renders correctly 1`] = `
<App
  loggedUser={[Function]}
  logout={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  user={null}
/>
`;
